### PR TITLE
Nested import ordering

### LINKED ISSUE #168: Incorrect nested import ordering
I just discovered this bug, where two imports that both import the same file can end up with the imported file out-of-order intermittently (due to parallel loading).

For example, if two separate files import `base.css`, in some cases the second file will process first and actually include it. Whereas the first file will not (due to de-duping) and then the result will be out-of-order and can affect the resulting applied styles (or variable definitions).

I have a full test case in my branch: https://github.com/adam-h/postcss-import/commit/569ea1ed4da1e7b84f1d190937a8341bbfac5b3b

Note how the first import is [delayed](https://github.com/adam-h/postcss-import/blob/569ea1ed4da1e7b84f1d190937a8341bbfac5b3b/test/order.js#L29-L31) so that the second will always process first (removing the randomness of occurance due to the filesystem).

### PR DESCRIPTION
**Pull request by RyanZim**
Had to edit a test that depended on the old, incorrect behavior.

Fixes #168 :tada:

### CODE REVIEW

**Comment by RyanZim** [comment]
This has a few problems, do not merge.

**Comment by RyanZim** [comment]
Fixed. @MoOx ?

File: index.js [hunk 1]
```diff
@@ -163,7 +164,7 @@ function parseStyles(
 ) {
   var statements = parseStatements(result, styles)
 
-  return Promise.all(statements.map(function(stmt) {
+  return Promise.resolve(statements).then(promiseEach(function(stmt) {
```
Comment (MoOx) [inline index.js]: what is the difference between promise each and Promise.all? Cause from [the doc](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Promise/all) I can see

>  The array of values maintains the order of the original iterable object, not the order that the promises were resolved in.

Comment (RyanZim) [inline index.js]: @MoOx I should have explained these changes better.

`map` runs the function down to `return resolveImportId()`. This starts a promise chain. `map` then goes on to the next item in `statements`, without waiting for `resolveImportId` to resolve. This causes concurrency. Depending on the order that the files are loaded, `state` is modified incorrectly. If `foo-second.css` loads first, `bar.css` is added to `state.importedFiles`. When `foo-first.css` is loaded, `bar.css` appears to be already loaded, however, it is in the incorrect order in the bundle.

`promiseEach` waits for `resolveImportId` to resolve before going to the next item in `statements`. promise-each is a standalone port of http://bluebirdjs.com/docs/api/promise.each.html.

This is a performance cut; but this is the quickest way to fix this. If someone can figure out a better deduping system that does not require serialized loading, great; I can't ATM.

Hope this is clear.

Comment (MoOx) [inline index.js]: Clear enough! Thanks!

**Comment by MoOx** [comment]
@RyanZim what is your npm nickname?

**Comment by MoOx** [comment]
You know have the full rights for this repo.

To cut a release you will need to:

- be logged on npm from your terminal
- update changelog to add a new section + bump version number in package.json (like explained here https://github.com/MoOx/npmpub#requirements, you don't need to install it, it's already a dev deps of this project)
- commit your changelog add + package.json update with a message being the version number for clarity
- run "npm run release". This should run test with a fresh node_modules (to ensure everything will work correctly from a fresh install), then will release a new version to npm and will cut a github release using notes from the changelog

If you are not confortable doing this, I can do this with you if you want, or I can even do it alone, it will take me a minute.

**Comment by RyanZim** [comment]
@MoOx ryanzim

https://www.npmjs.com/~ryanzim

**Comment by MoOx** [comment]
image

**Comment by RyanZim** [comment]
@MoOx Couple of questions:

- `master` requires the status check __continuous-integration/appveyor/branch__. Isn't that incorrect?
- Should this change be published immediately, or should we add a few other things first? CC: @ai

**Comment by MoOx** [comment]
I think when there are fork, continuous-integration/appveyor/branch is incorrectly not working as expected. No big deal imo.
You can release this asap as a patch. Do not hesitate to cut 3 releases in a day if you feel the need :) As soon as you respect semver (tldr: breaking change: major, new feature: minor, fix: patch).

**Comment by RyanZim** [comment]
> I think when there are fork, continuous-integration/appveyor/branch is incorrectly not working as expected.

Branch only runs on internal PRs.

**Comment by RyanZim** [comment]
@MoOx I would like to do releases in the future; but can you ship this today? I don't have time ATM.

**Comment by MoOx** [comment]
I just released 8.1.3

**Comment by RyanZim** [comment]
Thanks!

**Pull request merged**